File: Kotlin/151.reverse_words_in_a_string.kt
fun main() {
    println(reverseWords("the sky is blue"))
    println(reverseWords("  hello world  "))
    println(reverseWords("a good   example"))
}

fun reverseWords(s: String): String {
    val list = mutableListOf<String>()

    var temp = ""
    for (c in s) {
        if (temp.isEmpty() && c == ' ') {
            continue
        }

        if (c == ' ') {
            if (temp.isNotEmpty()) {
                list.add(temp)
            }

            temp = ""
        } else {
            temp += c
        }
    }

    if (temp.isNotEmpty() && temp.first() != ' ') {
        list.add(temp)
    }

    return list.reversed().joinToString(" ")
}

File: Dart/151.reverse_words_in_a_string.dart
void main() {
  print(reverseWords("the sky is blue"));
  print(reverseWords("  hello world  "));
  print(reverseWords("a good   example"));
}

String reverseWords(String s) {
  final list = <String>[];

  var temp = "";
  for (int i = 0; i < s.length; i++) {
    if (temp.isEmpty && s[i] == " ") {
      continue;
    }

    if (s[i] == " ") {
      if (temp.isNotEmpty) {
        list.add(temp);
      }

      temp = "";
    } else {
      temp += s[i];
    }
  }

  if (temp.isNotEmpty && temp[0] != " ") {
    list.add(temp);
  }

  return list.reversed.join(" ");
}
